File: backend/app/schemas/schedule.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# ---------- 配置 ----------
class ScheduleConfigBase(BaseModel):
    days: int = Field(2, ge=2, le=3)
    lanes: int = Field(6, ge=2, le=12)
    hard_rules: str = ""
    soft_rules: str = ""


class ScheduleConfigUpdate(ScheduleConfigBase):
    pass


class EntryUpdate(BaseModel):
    """赛次更新（只允许改开始/结束时间与场地，当天按时间排序）"""

    start_time: str = Field(default="", max_length=5)
    end_time: str = Field(default="", max_length=5)
    venue: str = Field(default="", max_length=100)


class EntryCreate(BaseModel):
    """新增赛次（空赛次，分组分道后续在详情里手动新增）"""

    event_id: int
    day_index: int = Field(ge=1, le=3)
    period: str = Field(pattern="^(上午|下午)$")
    round_type: str = Field(default="决赛", pattern="^(预赛|决赛)$")
    start_time: str = Field(default="", max_length=5)
    end_time: str = Field(default="", max_length=5)
    venue: str = Field(default="", max_length=100)


class LaneCreate(BaseModel):
    """在某组内新增一个分道/席位"""

    group_id: int
    athlete_id: int | None = None
    class_team_id: int | None = None


class ScheduleConfigOut(ScheduleConfigBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    academic_year_id: int
    ai_history: list[str] = Field(default_factory=list)
    generated_at: datetime | None = None


# ---------- 分道/席位 ----------
class LaneOut(BaseModel):
    id: int
    lane_no: int
    athlete_id: int | None = None
    class_team_id: int | None = None
    athlete_name: str | None = None
    number: int | None = None
    grade: str | None = None
    class_name: str | None = None
    result: str = ""
    rank: int | None = None


class GroupOut(BaseModel):
    id: int
    group_no: int
    lanes: list[LaneOut] = Field(default_factory=list)


# ---------- 赛次 ----------
class EntryOut(BaseModel):
    id: int
    event_id: int
    event_name: str
    group_name: str  # 组别(年级)
    gender: str
    is_team: bool
    day_index: int
    period: str
    round_type: str
    order_no: int
    group_count: int
    advance_count: int
    start_time: str
    end_time: str
    venue: str


class EntryDetail(EntryOut):
    groups: list[GroupOut] = Field(default_factory=list)


class ScheduleOut(BaseModel):
    """整表：配置 + 各半天分段的赛次列表。"""

    config: ScheduleConfigOut | None = None
    entries: list[EntryOut] = Field(default_factory=list)
    day_dates: dict[int, str] = Field(default_factory=dict)  # day_index -> 日期文案


# ---------- 成绩录入 ----------
class LaneResultIn(BaseModel):
    lane_id: int
    result: str = ""
    rank: int | None = None


class LaneUpdateIn(BaseModel):
    """手动修改分道选手"""

    lane_id: int
    athlete_id: int | None = None
    class_team_id: int | None = None


class ResultsUpdate(BaseModel):
    results: list[LaneResultIn] = Field(default_factory=list)


class AIOptimizeIn(BaseModel):
    # 对话框额外要求；可为空。为空时仅按规则+数据生成/优化
    message: str = Field(default="")


class ClearScheduleIn(BaseModel):
    admin_password: str = Field(..., min_length=1)


class AIOptimizeOut(BaseModel):
    """AI 生成/优化后返回的缺陷/优化报告"""

    issues: list[str] = Field(default_factory=list)
    mode: str = "optimize"  # "generate" 或 "optimize"

File: frontend/src/api/schedule.ts
import client from "./client";

export interface ScheduleConfig {
  id: number;
  academic_year_id: number;
  days: number;
  lanes: number;
  hard_rules: string;
  soft_rules: string;
  ai_history: string[];
  last_ai_request: string;
  last_ai_response: string;
  last_ai_report: string[];
  last_ai_mode: string;
  last_ai_time: string | null;
  generated_at: string | null;
}

export interface ScheduleConfigInput {
  lanes: number;
}

export interface EntryUpdateInput {
  start_time: string;
  end_time: string;
  venue: string;
}

export interface EntryCreateInput {
  event_id: number;
  day_index: number;
  period: string;
  round_type: string;
  start_time: string;
  end_time: string;
  venue: string;
}

export interface LaneCreateInput {
  group_id: number;
  athlete_id: number | null;
  class_team_id: number | null;
}

export interface ScheduleEntry {
  id: number;
  event_id: number;
  event_name: string;
  group_name: string;
  gender: string;
  is_team: boolean;
  day_index: number;
  period: string;
  round_type: string;
  order_no: number;
  group_count: number;
  advance_count: number;
  start_time: string;
  end_time: string;
  venue: string;
}

export interface ScheduleLane {
  id: number;
  lane_no: number;
  athlete_id: number | null;
  class_team_id: number | null;
  athlete_name: string | null;
  number: number | null;
  grade: string | null;
  class_name: string | null;
  result: string;
  rank: number | null;
}

export interface ScheduleGroup {
  id: number;
  group_no: number;
  lanes: ScheduleLane[];
}

export interface EntryDetail extends ScheduleEntry {
  groups: ScheduleGroup[];
}

export interface ScheduleData {
  config: ScheduleConfig | null;
  entries: ScheduleEntry[];
  day_dates: Record<string, string>;
}

export async function getConfig(): Promise<ScheduleConfig> {
  const { data } = await client.get<ScheduleConfig>("/schedule/config");
  return data;
}

export async function generateSchedule(): Promise<ScheduleData> {
  const { data } = await client.post<ScheduleData>("/schedule/generate");
  return data;
}

export async function updateConfig(payload: ScheduleConfigInput): Promise<ScheduleConfig> {
  const { data } = await client.put<ScheduleConfig>("/schedule/config", payload);
  return data;
}

export async function clearSchedule(adminPassword: string): Promise<ScheduleConfig> {
  const { data } = await client.post<ScheduleConfig>("/schedule/clear", {
    admin_password: adminPassword,
  });
  return data;
}

export interface AIOptimizeResult {
  issues: string[];
  mode: "generate" | "optimize";
}

export async function aiOptimize(message: string): Promise<AIOptimizeResult> {
  const { data } = await client.post<AIOptimizeResult>(
    "/schedule/ai-optimize",
    { message },
    { timeout: 0 }, // 生成/优化耗时较长，不设超时上限
  );
  return data;
}

export async function aiCheck(message: string): Promise<AIOptimizeResult> {
  const { data } = await client.post<AIOptimizeResult>(
    "/schedule/ai-check",
    { message },
    { timeout: 0 }, // AI 检查耗时较长，不设超时上限
  );
  return data;
}

export async function getSchedule(): Promise<ScheduleData> {
  const { data } = await client.get<ScheduleData>("/schedule");
  return data;
}

export async function updateEntry(entryId: number, input: EntryUpdateInput): Promise<ScheduleEntry> {
  const { data } = await client.put<ScheduleEntry>(`/schedule/entries/${entryId}`, input);
  return data;
}

export async function getEntryDetail(entryId: number): Promise<EntryDetail> {
  const { data } = await client.get<EntryDetail>(`/schedule/entries/${entryId}`);
  return data;
}

export async function createEntry(input: EntryCreateInput): Promise<ScheduleEntry> {
  const { data } = await client.post<ScheduleEntry>("/schedule/entries", input);
  return data;
}

export async function deleteEntry(entryId: number): Promise<void> {
  await client.delete(`/schedule/entries/${entryId}`);
}

export async function addGroup(entryId: number): Promise<EntryDetail> {
  const { data } = await client.post<EntryDetail>(`/schedule/entries/${entryId}/groups`);
  return data;
}

export async function deleteGroup(entryId: number, groupId: number): Promise<EntryDetail> {
  const { data } = await client.delete<EntryDetail>(`/schedule/entries/${entryId}/groups/${groupId}`);
  return data;
}

export async function addLane(entryId: number, input: LaneCreateInput): Promise<EntryDetail> {
  const { data } = await client.post<EntryDetail>(`/schedule/entries/${entryId}/lanes/add`, input);
  return data;
}

export async function deleteLane(entryId: number, laneId: number): Promise<EntryDetail> {
  const { data } = await client.delete<EntryDetail>(`/schedule/entries/${entryId}/lanes/${laneId}`);
  return data;
}

export interface LaneResult {
  lane_id: number;
  result: string;
  rank: number | null;
}

export interface LaneUpdate {
  lane_id: number;
  athlete_id: number | null;
  class_team_id: number | null;
}

export async function updateLanes(entryId: number, lanes: LaneUpdate[]): Promise<EntryDetail> {
  const { data } = await client.put<EntryDetail>(`/schedule/entries/${entryId}/lanes`, lanes);
  return data;
}

export async function updateResults(entryId: number, results: LaneResult[]): Promise<EntryDetail> {
  const { data } = await client.put<EntryDetail>(`/schedule/entries/${entryId}/results`, {
    results,
  });
  return data;
}

export async function buildFinals(entryId: number): Promise<EntryDetail> {
  const { data } = await client.post<EntryDetail>(`/schedule/entries/${entryId}/build-finals`);
  return data;
}
